pkg/controllers/systemd-etcd.go

```diff
@@ -0,0 +1,153 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/openshift/microshift/pkg/config"
+	"github.com/openshift/microshift/pkg/util/cryptomaterial"
+	"k8s.io/klog/v2"
+)
+
+type ExtEtcdService struct {
+	NodeName            string
+	DataDir             string
+	CertFile            string
+	KeyFile             string
+	TrustedCAFile       string
+	InitialUrl          string
+	PeerCertFile        string
+	PeerKeyFile         string
+	PeerTrustedCAFile   string
+	AdvertiseClientUrls string
+	ListenClientUrls    string
+	ListenPeerUrls      string
+	ListenMetricsUrls   string
+}
+
+func NewExtEtcd(cfg *config.MicroshiftConfig) *ExtEtcdService {
+	s := &ExtEtcdService{}
+	s.configure(cfg)
+	return s
+}
+
+func (s *ExtEtcdService) Name() string           { return "etcd" }
+func (s *ExtEtcdService) Dependencies() []string { return []string{} }
+
+func (s *ExtEtcdService) configure(cfg *config.MicroshiftConfig) {
+	s.NodeName = cfg.NodeName
+
+	s.DataDir = filepath.Join(microshiftDataDir, s.Name())
+	certsDir := cryptomaterial.CertsDirectory(microshiftDataDir)
+	etcdServingCertDir := cryptomaterial.EtcdServingCertDir(certsDir)
+	etcdPeerCertDir := cryptomaterial.EtcdPeerCertDir(certsDir)
+	etcdSignerCertPath := cryptomaterial.CACertPath(cryptomaterial.EtcdSignerDir(certsDir))
+
+	s.CertFile = cryptomaterial.PeerCertPath(etcdServingCertDir)
+	s.KeyFile = cryptomaterial.PeerKeyPath(etcdServingCertDir)
+	s.TrustedCAFile = etcdSignerCertPath
+
+	s.PeerCertFile = cryptomaterial.PeerCertPath(etcdPeerCertDir)
+	s.PeerKeyFile = cryptomaterial.PeerKeyPath(etcdPeerCertDir)
+	s.PeerTrustedCAFile = etcdSignerCertPath
+
+	s.InitialUrl = fmt.Sprintf("https://%s:2380", cfg.NodeIP)
+	s.AdvertiseClientUrls = fmt.Sprintf("https://%v:2379", cfg.NodeIP)
+	s.ListenClientUrls = fmt.Sprintf("https://127.0.0.1:2381,https://%v:2381", cfg.NodeIP)
+	s.ListenPeerUrls = fmt.Sprintf("https://%v:2380", cfg.NodeIP)
+	s.ListenMetricsUrls = "127.0.0.1:2381"
+}
+
+func (s *ExtEtcdService) Run(ctx context.Context, ready chan<- struct{}, stopped chan<- struct{}) error {
+	defer close(stopped)
+
+	// Check to see if we should run as a systemd run or directly as a binary.
+	runningAsSvc := false
+	if os.Getenv("INVOCATION_ID") != "" {
+		runningAsSvc = true
+	}
+
+	execPath, err := os.Executable()
+	if err != nil {
+		return fmt.Errorf("%v failed to get exec path: %v", s.Name(), err)
+	}
+
+	var etcdPath string
+	var args []string
+	// This is a temporary measure to run embedded etcd with the transient systemd unit.
+	// TODO remove this once we have the etcd binary laid down.
+	if os.Getenv("EXTERNAL_ETCD") == "" {
```

Comment:
What I did in my poc was using `INVOCATION_ID` env var - if MicroShift is running as systemd service - it'll be set